Blog - ucacher (#862)
earthly/website: blog/_posts/2024-12-02-ucacher.md

---

> title: "ucacher: Speeding up GitHub Actions via syscall instrumentation"

Reviewer: nit: I get that the CLI is lowercase `ucacher` but kind of think we should treat it as a proper noun in writing i.e. `Ucacher`.

toc: true
author: Ignacio del Valle Alles

internal-links:
 - Github Actions
 - caching
 - syscall
 - ptrace
topcta: false
excerpt: |
    This post introduces ucacher: a powerful new tool from Earthly that simplifies file caching and command skipping in GitHub Actions. Designed for effortless adoption and maximum precision, ucacher eliminates manual errors and supports a broad range of use cases.
categories:
  - githubactions
---

> *Introducing [ucacher](https://github.com/earthly/setup-ucacher): a powerful new tool from Earthly that simplifies file caching and command skipping in GitHub Actions. Designed for effortless adoption and maximum precision, `ucacher` eliminates manual errors and supports a broad range of use cases.*

Reviewer: I feel like "precision" is maybe not our strong suit 😅 like we aren't perfectly precise at detecting what should be a cache hits or missn - maybe we could say something like "Designed to significantly reduce build times in CI with little-to-no adoption effort." ?

Reviewer: Also what do we mean by " `ucacher` eliminates manual errors" ?

Author: precision is explained later on, it refers to whether a file is used or not by the command. It is true that the "maximum" can be relaxed :)

# `ucacher`: Speeding up GitHub Actions via syscall instrumentation

Traditional CI/CD workflows, particularly in GitHub Actions, rely on manually configured caching and skipping techniques to reduce redundancy and improve efficiency. While effective in basic scenarios, these methods often fall short in handling complex workflows, requiring developers to manage cache keys, paths, and conditional logic manually prone to errors and inefficiencies.

> Enter `ucacher`, a groundbreaking tool from Earthly that revolutionizes this process. Unlike traditional approaches, `ucacher` automates caching and skipping at the command level, eliminating the need for manual configuration and achieving unparalleled precision. With `ucacher`, developers can skip redundant tasks seamlessly, optimize resource use, and dramatically reduce build times, transforming the way CI/CD pipelines operate.

Reviewer: > Enter `ucacher`, a groundbreaking tool from Earthly that revolutionizes this process.

another nit, but  I feel like this line sounds a bit too sales-y for a technical audience (which I assume is the target for this, right?)

Author: yes, I was thinking the same

> ## How it's done today: the manual approach

Reviewer: Nit: I feel like we spend a lot of time as a tutorial for Github actions default caching/skipping here - I wonder if we should shorten this section a bit to focus more on ucacher? The readers may also be familiar enough with these

Author: good point, I will try to shorten it

#### ptrace

[ptrace(2)](https://man7.org/linux/man-pages/man2/ptrace.2.html) is a special syscall that allows a process (typically a debugger) to observe and control the execution of another process. It provides powerful functionality for debugging and instrumentation, such as inspecting and modifying memory, registers, and system calls of a target process.

#### The idea

`ucacher` launches the command and monitors the file-related syscalls triggered from its process tree by using `ptrace`.

When a file is open it halts the process execution and registers the file access. Depending on the file mode, `ucacher` tags the file as an input or output file. For input files it computes their hash, to capture the initial state.

After the execution completes successfully it uploads output files to persistent storage (e.g., GitHub Actions cache) along with some build metadata, in particular the hashes of all input files at the time they were read.

On subsequent runs, `ucacher` checks for matching initial conditions (input files content, arguments, environment variables, system architecture, etc.). If they match, it skips execution and restores the cached output files instead.

> ## Preliminary results

Author: I would leave that for a follow up post, once we opensource it

> ## Preliminary results

Author: added a reference to it

`ucacher` has already proven to be highly effective at reducing overall computation costs, especially in:

* Workflows with task sharding, such as compilation and testing, where workloads are distributed across multiple shards.
* Monorepos, where changes often have a limited scope, allowing for precise caching and skipping.

## Summary

In summary, `ucacher` simplifies and optimizes CI/CD workflows by making caching and skipping smarter and more precise:

* **Higher precision**: Tracks exact file dependencies, skipping commands only when truly unaffected.
* **Automation**: Eliminates manual cache keys and reduces configuration errors.
* **Efficiency**: Minimizes redundant steps, saving time and reducing compute costs.
* **Handles complexity**: Excels in large repositories, matrix builds, and multi-stage pipelines.
* **Seamless integration**: Works out of the box with tools like GitHub Actions.

> We’re eager to share `ucacher` with the developer community and would love to hear your feedback. Whether it’s suggestions, feature ideas, or insights from your own use cases, your input is essential to helping us improve and evolve it.

Reviewer: Link to github repo here maybe?

In this post, we’ll explore the current standard for implementing caching and skipping in GitHub Actions, to later on introduce Ucacher, a tool we’ve recently developed to automate these tasks at the command level, removing the need for manual setup and delivering greater precision.

## How it's done today: the manual approach

<div class="notice--info">
This section describes how skipping and caching are typically implemented in Github Actions. Feel free to [skip it](#the-problem-broad-cache-keys) ;) if you are familiarized with these techniques.
</div>

Skipping and caching are two techniques used in CI/CD pipelines to avoid redundant operations and reduce execution, which translates into faster iteration loops for devs and reduced cloud infrastructure costs.

### Caching
Caching involves storing files or directories in a persistent storage location so that they can be retrieved and reused in future executions of the workflows.

> [actions/cache](https://github.com/actions/cache) is the official Github action for caching files. It works by caching a set of directories or files specified by the path parameter, and identifying their state by a unique key.

Reviewer 2: ```suggestion
[actions/cache](https://github.com/actions/cache) is the official GitHub action for caching files. It works by caching a set of directories or files specified by the path parameter, and identifying their state by a unique key.
```

      restore-keys: ${{ '{{' }} runner.os {{ '}}' }}-node-
  
  - name: Step 3 - Install dependencies
    run: npm install
  
  - name: Step 4 - Build project
    run: npm run build
~~~

Restoring `./node_modules` from the cache will significantly reduce the runtime of `npm install`, since all the dependencies will be found locally.

### Skipping

Skipping ensures that unnecessary steps are not executed, based on certain conditions. This reduces runtime and resource consumption.
In GitHub Actions, [conditional job runs](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/using-conditions-to-control-job-execution) allow you to control whether a job or step is executed based on specific conditions using the `if` keyword.
This enables dynamic workflows, optimizing execution by skipping unnecessary jobs or steps based on [context](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs) information.

Combining caching and skipping in GitHub Actions is typically done by:
1. First caching
2. Then skipping if a cache hit was found

> Here it is how the previous example can be improved to completely skip step 3a, when there is cache hit on step 2a:

Reviewer 2: ```suggestion
Here is how the previous example can be improved to completely skip step 3a, when there is cache hit on step 2a:
```

  - name: Step 2b - Restore build
    id: cache-build
    uses: actions/cache@v3
    with:
      path: ./build
      key: ${{ '{{' }} runner.os {{ '}}' }}-node-build-${{ '{{' }} hashFiles('**/package-lock.json') {{ '}}' }}-${{ '{{' }} hashFiles('src/**') {{ '}}' }}
  
  - name: Step 3b - Conditionally restore dependencies
    id: cache-dependencies
    if: steps.cache-build.outputs.cache-hit != 'true'
    uses: actions/cache@v3
    with:
      path: ./node_modules # Directory to cache
      key: ${{ '{{' }} runner.os {{ '}}' }}-node-${{ '{{' }} hashFiles('**/package-lock.json') {{ '}}' }}
  
  - name: Step 4b - Conditionally install dependencies
    if: steps.cache-build.outputs.cache-hit != 'true' && 
        steps.cache-dependencies.outputs.cache-hit != 'true'
    run: npm install
  
  - name: Step 5b - Build project
    if: steps.cache-build.outputs.cache-hit != 'true'
    run: npm run build
~~~ 

As you can see here, we’ve changed the `path` parameter to `./build`, since that is the folder that the command generates, and also augmented the key to also include the hashes of all files under `src`, since all of them might potentially be used by the build command.

> This example now presents the advantage that the build step is also skipped when the workflow runs for a second time over the same package-lock.json and./src folder contents, for example when you change your `README.md` file or when you rollback a commit. However, notice how the workflow readability has degraded.  

Reviewer 2: ```suggestion
This example now presents the advantage that the build step is also skipped when the workflow runs for a second time over the same package-lock.json and ./src folder contents, for example when you change your `README.md` file or when you rollback a commit. However, notice how the workflow readability has degraded.  
```

## The problem: Broad cache keys

Now, let’s explore a more advanced scenario where the job uses a matrix strategy to divide test execution across multiple parallel runners:

~~~{.yml caption=""}
build:
  name: yarn build and lint
  runs-on: ubuntu-latest
  strategy:
    matrix:
      shard:
      - 1/5
      - 2/5
      - 3/5
      - 4/5
      - 5/5
  steps:
  - name: Step 1c - Checkout code
    uses: actions/checkout@v3

  - name: Step 2c - Restore dependencies
    id: cache-dependencies
    uses: actions/cache@v3
    with:
      path: ./node_modules # Directory to cache
      key: ${{ '{{' }} runner.os {{ '}}' }}-node-${{ '{{' }} hashFiles('**/package-lock.json') {{ '}}' }}
  
  - name: Step 3c - Conditionally install dependencies
    if: steps.cache-dependencies.outputs.cache-hit != 'true'
    run: npm install
  
  - name: Step 4c - Run tests
    run: npm run test --shard=${{ '{{' }} matrix.shard {{ '}}' }} --coverage
~~~ 

> This is how the previous job definition works: at runtime, it creates 5 different runners (one per matrix element), and each runner runs a job with a different `matrix.shard `value.

Reviewer 2: ```suggestion
This is how the previous job definition works: at runtime, it creates 5 different runners (one per matrix element), and each runner runs a job with a different `matrix.shard` value.
```

It improves the user experience offered by `actions/cache` in the following ways:
* No parametrization is required
  * No human errors involved
  * No prior knowledge of the command to cache is required
* Finer grain caching: at the command level, vs to the step level.
* Greater precision: If the modified files triggering the build don’t affect the command it will be skipped
* Automatic output caching: It restores the files written by the command in case of a cache hit (skip)

This is how the previous example could be cached with Ucacher:

~~~{.yml caption=""}
build:
  name: yarn build and lint
  runs-on: ubuntu-latest
  strategy:
    matrix:
      shard:
      - 1/5
      - 2/5
      - 3/5
      - 4/5
      - 5/5
  steps:
  - name: Step 1e - Checkout code
    uses: actions/checkout@v3
  
  - name: Step 2e - Install ucacher
  - uses: earthly/setup-ucacher@main
  
  - name: Step 3e - Conditionally install dependencies
    run: ucacher npm install
  
  - name: Step 4e - Conditionally run tests
    run: ucacher npm run test --shard=${{ '{{' }} matrix.shard {{ '}}' }} --coverage
~~~ 

> For example, suppose now that you push a change in a test file belonging to the 3rd shard, then Ucacher automatically detects this file was not used in any previous runs for the shards 1,2,4,5 and skip them all, and only runs the tests for the 3rd shard.

Reviewer 2: ```suggestion
For example, suppose now that you push a change in a test file belonging to the 3rd shard, then Ucacher automatically detects this file was not used in any previous runs for the shards 1,2,4,5 and skips them all, and only runs the tests for the 3rd shard.
```

[ptrace(2)](https://man7.org/linux/man-pages/man2/ptrace.2.html) is a special syscall that allows a process (typically a debugger) to observe and control the execution of another process. It provides powerful functionality for debugging and instrumentation, such as inspecting and modifying memory, registers, and system calls of a target process.

#### The idea

Ucacher launches the command and monitors the file-related syscalls triggered from its process tree by using `ptrace`. 

When a file is open it halts the process execution and registers the file access. Depending on the file mode, Ucacher tags the file as an input or output file. For input files it computes their hash, to capture the initial state.

After the execution completes successfully, it uploads output files to persistent storage (e.g., GitHub Actions cache) along with some build metadata, in particular the hashes of all input files at the time they were read.

On subsequent runs, Ucacher checks for matching initial conditions (input files content, arguments, environment variables, system architecture, etc.). If they match, it skips execution and restores the cached output files instead.

#### The insight

> While it’s impossible to know in advance which files an arbitrary command will access during execution, Ucacher is based on the observation that given a past execution, we can check if its initial state is maintained for the actual running environment to conclude that a new execution would produce the same results. 

Reviewer 2: ```suggestion
While it’s impossible to know in advance which files an arbitrary command will access during execution, Ucacher is based on the observation that given a past execution, we can check if its initial state is maintained to the extent that the running environment can conclude that a new execution would produce the same results. 
```

Author: not exactly what I was trying to express. Please check this new version

Reviewer 2: New version is clearer. Thanks.

As a consequence, the approach to identifying matching initial conditions differs from the traditional one-hop cache lookup. Ucacher employs a two-phase lookup process instead:
1. First, retrieve all cached entries associated with the specific combination of: command, environment, operating system, and branch.
2. Then, for each potential entry, verify whether the hashes of the input files from that execution match the current filesystem state. If they match, then we have a cache hit. 

## Preliminary results

> To evaluate the real-world effectiveness of Ucacher, we tested it on the well-known [facebook/react](https://github.com/facebook/react) repository. With its complexity and already highly optimized workflows, it provided an excellent benchmark for assessing its capabilities.

Reviewer 2: ```suggestion
To evaluate the real-world effectiveness of Ucacher, we tested it on the well-known [facebook/react](https://github.com/facebook/react) repository. With its complexity and already highly optimized workflows, it provided an excellent benchmark for assessing Ucacher's capabilities.
```

Ucacher has already proven to be highly effective at reducing overall computation costs, especially in:

* Workflows with task sharding, such as compilation and testing, where workloads are distributed across multiple shards.
* Monorepos, where changes often have a limited scope, allowing for precise caching and skipping.

## Summary

In summary, Ucacher simplifies and optimizes CI/CD workflows by making caching and skipping smarter and more precise:

* **Higher precision**: Tracks exact file dependencies, skipping commands only when truly unaffected.
* **Automation**: Eliminates manual cache keys and reduces configuration errors.
* **Efficiency**: Minimizes redundant steps, saving time and reducing compute costs.
* **Handles complexity**: Excels in large repositories, matrix builds, and multi-stage pipelines.
* **Seamless integration**: Works out of the box with tools like GitHub Actions.

We're eager to share Ucacher with the developer community and would love to hear your feedback. Whether it's suggestions, feature ideas, or insights from your own use cases, your input is essential to helping us improve and evolve it.

Give it a try and let us know how it works for you!

> <https://github.com/earthly/setup-ucacher>

Reviewer 3: We generally avoid having bare links in the body of the article.

```suggestion
[Visit setup-ucacher on GitHub to get started!](https://github.com/earthly/setup-ucacher)
```
